### PR TITLE
feat(meta): add checkpoint_frequency and support decoupling

### PR DESCRIPTION
**Pull request by xxhZs**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Support decoupling barrier and checkpoint. After this pr, we will have two kinds of barriers. We will inject a barrier every barrier_interval_ms ms, and Inject one checkpoint per checkpoint_frequency barrier
The data we are currently reading is still done checkpoint. we need to do some work in Frontend to support reading data without checkpoint.
We add a parameter(checkpoint) to flush. If it is false, It will not be guaranteed that the visible data has been checkpoint

**It takes a lot of time to run `datagen.slt` and `nexmark_endless.slt` because we need to wait for the checkpoint to finish creating mv. So I add flush.**

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by hzxa21** [comment]
Let's hold this PR until we figure out the cause of #4861.

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:48]:
```suggestion
    pub fn notify_checkpoint_barrier_collected(&mut self) {
```

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:56]:
```suggestion
    pub fn notify_barrier_collected(&mut self) {
```

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:32]: Will this be triggered if barrier(checkpoint = true) is collected? If yes, I suggest rename the two newly added fields to collected_checkpoint_barrier and collected_barrier.

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:62]:
```suggestion
    pub fn need_collect_checkpoint_barrier(&self) -> bool {
```

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:68]:
```suggestion
    pub fn notify_checkpoint_barrier_collection_failed(mut self, err: MetaError) {
```

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [inline src/meta/src/barrier/notifier.rs:77]:
```suggestion
    pub fn notify_barrier_collection_failed(&mut self, err: MetaError) {
```

File: src/meta/src/barrier/notifier.rs
Comment (hzxa21) [reply src/meta/src/barrier/notifier.rs:32]: and is it better to use a enum to wrap these two types of sender?

File: src/meta/src/barrier/mod.rs
Comment (xxhZs) [reply src/meta/src/barrier/mod.rs:342]: Don't have a problem. If we make a mistake between these two steps, we will do recovery and remove all changes in mem. It is like removing it before the state is committed to meta store

File: src/meta/src/barrier/mod.rs
Comment (hzxa21) [inline src/meta/src/barrier/mod.rs:881]:
```suggestion                    
                    assert_eq!(checkpoint, notifier.bound_to_checkpoint_barrier())                   
```

**Review by hzxa21 — APPROVED** [review]
Rest LGTM.

Please also run a longevity test before merging.

**Comment by BugenZhao** [comment]
I'd also like to take a look. 🥰

File: src/meta/src/barrier/mod.rs
Comment (BugenZhao) [inline src/meta/src/barrier/mod.rs:600]:
```suggestion
            let checkpoint = checkpoint || checkpoint_control.try_get_checkpoint();
```

Comment (BugenZhao) [reply src/meta/src/barrier/mod.rs:600]: I think it should be the `ScheduledBarriers`'s responsibility to decide whether a plain barrier should be a checkpoint, as the periodic barrier is generated by it. IIUC, `CheckpointControl` should be used to manage the **in-flight** barriers.

File: src/meta/src/barrier/mod.rs
Comment (BugenZhao) [inline src/meta/src/barrier/mod.rs:753]:
```suggestion
                fail_node,
```

File: src/meta/src/barrier/progress.rs
Comment (BugenZhao) [inline src/meta/src/barrier/progress.rs:153]: Should still keep `unreachable` here?

File: src/meta/src/barrier/schedule.rs
Comment (BugenZhao) [inline src/meta/src/barrier/schedule.rs:99]:
```suggestion
                    notifiers: new_notifiers,
```

File: src/meta/src/barrier/schedule.rs
Comment (BugenZhao) [inline src/meta/src/barrier/schedule.rs:97]:
```suggestion
                // If no command scheduled, create a periodic barrier by default.
```

File: src/meta/src/barrier/schedule.rs
Comment (BugenZhao) [inline src/meta/src/barrier/schedule.rs:86]: I think it's kind of weird to add a flag of `checkpoint` in the notifier. For now, the `checkpoint` flags in the notifiers seem redundant to `Scheduled::checkpoint`. To tell that we should flush with checkpoints, I think we may simple add a flag to the parameters of this function, and there's no need to store it anywhere.

File: src/meta/src/barrier/schedule.rs
Comment (BugenZhao) [inline src/meta/src/barrier/schedule.rs:100]: I named the periodic one with `checkpoint` as I was to cover the checkpoint flag inside the command. If we decide to put it in `Scheduled`, I think we'd better give a name other than `checkpoint` to this function.

File: src/meta/src/barrier/mod.rs
Comment (BugenZhao) [inline src/meta/src/barrier/mod.rs:897]: So creating an empty materialized view will last for 2 barrier intervals in most cases? 🤣 This may make the e2e tests much slower in CI. How about decreasing the barrier interval?

File: src/stream/src/task/barrier_manager/managed_state.rs
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:50]:
```suggestion
    inner: ManagedBarrierStateInner,
```

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM. Great job! 🚀 I have some comments on the code structure, please check the review for details.

File: src/meta/src/barrier/schedule.rs
Comment (BugenZhao) [inline src/meta/src/barrier/schedule.rs:154]: It seems unnecessary to always be `true` here. For commands like dropping materialized views, rescheduling, or `resume`, a normal barrier should be enough.

I still suggest implying the `is_checkpoint` in `Command` itself so that we can control it more preciously. For plain commands, we may use a bool flag to represent it.

File: src/frontend/src/handler/query.rs
Comment (fuyufjh) [reply src/frontend/src/handler/query.rs:205]: Sounds good to me

File: src/meta/src/barrier/mod.rs
Comment (xxhZs) [reply src/meta/src/barrier/mod.rs:881]: If checkpoint is true, `bound_to_checkpoint_barrier()` may be false.

File: src/meta/src/barrier/mod.rs
Comment (BugenZhao) [reply src/meta/src/barrier/mod.rs:897]: e2e compared to other PRs. 🥵
image
image

Comment (fuyufjh) [reply src/meta/src/barrier/mod.rs:897]: But barrier intervals can be further decreased?

Comment (xxhZs) [reply src/meta/src/barrier/mod.rs:897]: I have found other pr's that also take more than 9 minutes to run. So maybe this is not the main reason

File: src/stream/src/task/barrier_manager/managed_state.rs
Comment (xxhZs) [reply src/stream/src/task/barrier_manager/managed_state.rs:149]: yes, because we are now sorted to proceed `sealed_epoch` in `managed_state`. So we make sure that all the ones less than or equal to`barrier.prev_epoch` are cleared when we do `stop_all_actors`. If not, We will be blocked in `collect_barrier`

Comment (xxhZs) [reply src/stream/src/task/barrier_manager/managed_state.rs]: we should use `>`, And I use `drain_collect_rx` before `send_barrier` in `stop_all_actors` to avoid blocking `collect_barrier` calls

**Comment by xxhZs** [comment]
localhost_54426_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s from=now-1h to=now
main and decouple benchmark

**Comment by xxhZs** [comment]
https://github.com/Mergifyio requeue

**Review by BugenZhao — APPROVED** [review]
LGTM!

File: src/meta/src/barrier/mod.rs
Comment (BugenZhao) [reply src/meta/src/barrier/mod.rs]: How is this resolved now?

**Pull request merged**
